### ISSUE TITLE
patchelf gets confused when there is more than one `.shstrtab` section

### ISSUE THREAD

**Issue by vapier** [NixOS/patchelf#197]
don't ask me why the ELF is the way it is, i'm just the messenger :)

Centos 6's glibc ships a libcrypt-2.12.so that has two `.shstrtab` sections.  when patchelf runs `sortShdrs()` on this file to update `e_shstrndx`, it picks the wrong one, and this corrupts the section table parsing.

runtime doesn't care because section headers are purely used by offline debugging, not at runtime.  but developer tools (like `readelf`) generally get upset :).

to reproduce:
```sh
$ wget http://mirror.centos.org/centos/6/os/x86_64/Packages/glibc-2.12-1.212.el6.x86_64.rpm
$ rpm2tar glibc-2.12-1.212.el6.x86_64.rpm
$ tar xvf glibc-2.12-1.212.el6.x86_64.tar ./lib64/libcrypt-2.12.so
$ readelf -WSh ./lib64/libcrypt-2.12.so
  Section header string table index: 33
  [29] .shstrtab         STRTAB          0000000000000000 0080c5 000198 00      0   0  1
  [33] .shstrtab         STRTAB          0000000000000000 009600 000123 00      0   0  1
$ cp ./lib64/libcrypt-2.12.so ./test.elf
$ patchelf --debug --set-soname libcrypt-2-a5d5a997.12.so test.elf
rewriting symbol table section 29     <-- oops!
$ readelf -WSh test.elf
  Section header string table index: 28    <-- bad!
  [28] b                 STRTAB          0000000000000000 0080c5 000198 00      0   0  1
  [30] .shstrtab         STRTAB          0000000000000000 008fb0 000632 00      0   0  1
  [31] .version_r        PROGBITS        0000000000000000 0095e4 00001c 00      0   0  4
  [32] b                 STRTAB          0000000000000000 009600 000123 00      0   0  1
  [33] s                 STRTAB          0000000000237230 237230 0001ae 00   A  0   0  8    <-- corrupted section names!
```

i think this is what's causing trouble for the Python manylinux2010 Centos 6 docker container as reported at https://github.com/pypa/manylinux/issues/411.

**Comment by domenkozar** [comment]
What version of patchelf and any additional patches used?

**Comment by domenkozar** [comment]
With the latest master I get:

```
$ /nix/store/d760wxwkslrdx8wvx6b0jsir865729qj-patchelf-0.10.1234.abcdef/bin/patchelf --debug --set-soname libcrypt-2-a5d5a997.12.so test.elf
patching ELF file 'test.elf'
Kernel page size is 4096 bytes
new SONAME is 'libcrypt-2-a5d5a997.12.so'
SONAME is too long, resizing...
replacing section '.dynstr' with size 430
this is a dynamic library
last page is 0x237000
needed space is 504
rewriting section '.dynstr' from offset 0x5e0 (size 404) to offset 0xa000 (size 430)
rewriting section '.note.ABI-tag' from offset 0x25c (size 32) to offset 0xa1b0 (size 32)
rewriting section '.note.gnu.build-id' from offset 0x238 (size 36) to offset 0xa1d0 (size 36)
rewriting symbol table section 2
rewriting symbol table section 27
writing test.elf
```

**Comment by vapier** [comment]
sorry, i should have mentioned versions i tested against.  i noticed the bug in the version shipped by the manylinux project (which is 0.10 i think), and then i switched over to using git directly.  2ba64817ec6f3b714503ea6e6aa8439505bb7393 was my base for figuring out a hack to mitigate things.

**Comment by pablogsal** [comment]
This will be solved once #241 is merged

**Comment by pablogsal** [comment]
See also #240
